addons/sale/wizard/sale_order_discount.py

```diff
@@ -99,66 +91,47 @@ def _get_discount_product(self):
-
-        discount_dp = self.env['decimal.precision'].precision_get('Discount')
-        if not total_price_per_tax_groups:
-            # No valid lines on which the discount can be applied
-            return
-        if len(total_price_per_tax_groups) == 1:
-            # No taxes, or all lines have the exact same taxes
-            taxes = next(iter(total_price_per_tax_groups.keys()))
-            subtotal = total_price_per_tax_groups[taxes]
-            vals_list = [{
-                **self._prepare_discount_line_values(
-                    product=discount_product,
-                    amount=subtotal * discount_percentage,
-                    taxes=taxes,
-                    description=_(
-                        "Discount %(percent)s%%",
-                        percent=float_repr(discount_percentage * 100, discount_dp),
-                    ),
-                ),
-            }]
-        else:
-            vals_list = [
-                self._prepare_discount_line_values(
-                    product=discount_product,
-                    amount=subtotal * discount_percentage,
-                    taxes=taxes,
-                    description=_(
-                        "Discount %(percent)s%%"
-                        "- On products with the following taxes %(taxes)s",
-                        percent=float_repr(discount_percentage * 100, discount_dp),
-                        taxes=", ".join(taxes.mapped('name')),
-                    ) if self.discount_type != 'amount' else _(
-                        "Discount"
-                        "- On products with the following taxes %(taxes)s",
-                        taxes=", ".join(taxes.mapped('name')),
-                    )
-                ) for taxes, subtotal in total_price_per_tax_groups.items()
-            ]
-        return self.env['sale.order.line'].create(vals_list)
+        if self.discount_type == 'so_discount':
+            amount_type = 'percent'
+            amount = self.discount_percentage * 100.0
+            so_line_description = _("Discount %(percent)s%%", percent=amount)
+        else:  # self.discount_type == 'amount':
+            amount_type = 'fixed'
+            amount = self.discount_amount
+            so_line_description = _("Discount")
+
+        order = self.sale_order_id
+        AccountTax = self.env['account.tax']
+        order_lines = order.order_line.filtered(lambda x: not x.display_type)
+        base_lines = [
+            line._prepare_base_line_for_taxes_computation(
+                product_id=discount_product,
+                name=so_line_description,
+            )
+            for line in order_lines
+        ]
+        AccountTax._add_tax_details_in_base_lines(base_lines, order.company_id)
+        AccountTax._round_base_lines_tax_details(base_lines, order.company_id)
+
+        global_discount_base_lines = AccountTax._prepare_global_discount_lines(
```

Comment:
But you'll lose the name/description here anyway. You don't keep values like name in that one.